opt: fix row count estimates for inverted filters

mgartner opened the pull request:
#### opt: assign table ID to hypothetical inverted index columns

This commit fixes a minor bug in index recommendations where
hypothetical inverted index columns were not assigned table IDs in the
hypothetical table's metadata. In theory this could cause internal
errors, however, there are no known reproductions of errors.

Release note: None

#### opt: decrease row count rounding precision

Row counts are now rounded with less precision so that optimizer test
outputs are the same across different CPU architectures. Follow-up
of #75723.

Release note: None

#### opt: fix row count estimates for inverted filters

Previously, inverted filter row counts were set to their input's row
count. This was often an over-estimation because inverted filters
de-duplicate inverted index tuples from their input by primary key, and
inverted indexes can contain multiple tuples for each logical row.

This commit fixes this over-estimation by setting the row count of
inverted filters to the distinct count of the primary key columns from
the input.

Fixes #120055

Release note (performance improvement): A bug has been fixed that caused
the optimizer to over-estimate the cost of inverted index scans in some
cases. Now, plans with inverted index scans should be selected in more
cases where they are optimal.

mgartner committed:
opt: fix row count estimates for inverted filters

Previously, inverted filter row counts were set to their input's row
count. This was often an over-estimation because inverted filters
de-duplicate inverted index tuples from their input by primary key, and
inverted indexes can contain multiple tuples for each logical row.

This commit fixes this over-estimation by setting the row count of
inverted filters to the distinct count of the primary key columns from
the input.

Fixes #120055

Release note (performance improvement): A bug has been fixed that caused
the optimizer to over-estimate the cost of inverted index scans in some
cases. Now, plans with inverted index scans should be selected in more
cases where they are optimal.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120079)
<!-- Reviewable:end -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice fix!

> Reviewed 1 of 1 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120079#-:-NsQ0UAR-22AI1yNWleq:b-46lt7k)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice!

> Reviewed 1 of 1 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120079#-:-NsnmdCp0A_3EsRVchL1:b-8gk8uj)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+

yuzefovich commented:
Looks like some tests need an adjustment.

bors r-

michae2 commented:
Does this need to be backported?

mgartner commented:
> Does this need to be backported?

No. If we did, we'd want a session setting to avoid changing query plans unexpectedly.

mgartner commented:
bors r+

yuzefovich commented:
Should we backport this to 24.1?

yuzefovich commented:
Looks like `GitHub Actions Essential CI / unit_tests (pull_request)` part of CI is red.

bors r-

mgartner committed:
opt: assign table ID to hypothetical inverted index columns

This commit fixes a minor bug in index recommendations where
hypothetical inverted index columns were not assigned table IDs in the
hypothetical table's metadata. In theory this could cause internal
errors, however, there are no known reproductions of errors.

Release note: None

mgartner commented:
Added a new commit to fix a minor bug that this uncovered. PTAL.

mgartner commented:
Failure is really hard to find in the test output:

```        /var/lib/engflow/.../pkg/sql/opt/memo/testdata/stats/inverted-geo:142:
         SELECT i FROM t WHERE st_intersects('LINESTRING(100 100, 150 150)', g) ORDER BY i LIMIT 1
        output didn't match expected:
        @@ -19,24 +19,24 @@
               │    ├── stats: [rows=222.2222, distinct(2)=7, null(2)=0]
               │    ├── ordering: +1
               │    ├── limit hint: 1.00
               │    ├── sort
               │    │    ├── columns: i:1(int) g:2(geometry)
        -      │    │    ├── stats: [rows=4e-07]
        +      │    │    ├── stats: [rows=4.000001e-07]
               │    │    ├── ordering: +1
               │    │    ├── limit hint: 0.00
               │    │    └── index-join t
               │    │         ├── columns: i:1(int) g:2(geometry)
        -      │    │         ├── stats: [rows=4e-07]
        +      │    │         ├── stats: [rows=4.000001e-07]
               │    │         └── inverted-filter
               │    │              ├── columns: rowid:3(int!null)
               │    │              ├── inverted expression: /6
               │    │              │    ├── tight: false, unique: false
               │    │              │    └── union spans: ["B\xfd\xff\xff\xff\xff\xff\xff\xff\xff", "B\xfd\xff\xff\xff\xff\xff\xff\xff\xff"]
               │    │              ├── pre-filterer expression
               │    │              │    └── st_intersects('010200000002000000000000000000594000000000000059400000000000C062400000000000C06240', g:2) [type=bool]
        -      │    │              ├── stats: [rows=4e-07]
        +      │    │              ├── stats: [rows=4.000001e-07]
               │    │              ├── key: (3)
               │    │              └── scan t@t_g_idx

```

mgartner commented:
@mgartner ping

yuzefovich commented:
@mgartner friendly ping 😃

mgartner commented:
Unfortunately, this isn't a full solution. See more details here: https://github.com/cockroachdb/cockroach/issues/120055#issuecomment-1984632867

Maybe I should break out the first commit though?

yuzefovich commented:
Hm, I see. It seems useful to merge at least the first commit.

DrewKimball commented:
I'm confused by that comment. Either the index join is a part of the original memo group for which the inverted scan plan is being generated, in which case in inherits that group's stats. Otherwise, it's part of a new memo group that uses the row count estimate for the inverted filter to determine its own row count, so this change should apply. Am I missing something?

mgartner commented:
>I'm confused by that comment. Either the index join is a part of the original memo group for which the inverted scan plan is being generated, in which case in inherits that group's stats. Otherwise, it's part of a new memo group that uses the row count estimate for the inverted filter to determine its own row count, so this change should apply. Am I missing something?

I'm confused too... I don't remember exactly what I meant by that comment. My best guess is that I was highlighting that in the case that the index-join is part of the original memo group, it's cost may still be higher than it should because index-join cost is highly dependent on output row count and that row count could be higher than it should because it has the same row count as the original expression in the group.

I don't see any reason not to merge this, so I've rebased and I will bors when it is green.

mgartner committed:
opt: decrease row count rounding precision

Row counts are now rounded with less precision so that optimizer test
outputs are the same across different CPU architectures. Follow-up
of #75723.

Release note: None

mgartner commented:
I had to add a new commit to decrease the precision of the formatted `row-count` so that it matches on ARM and AMD64. I should probable backport that commit to active release branches to make other backports less painful.

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r5, 7 of 7 files at r6, 9 of 9 files at r7, 61 of 61 files at r8, 7 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120079#-:-OMh-dqq8-bcEYMqPtgI:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+
